* Curriculum Number: CSIE5137
* Course Name: Web Retrieval and Mining
* Student ID: r13921031
* Subject: Assignment 2
* The notebook can be accessed through this [link]()

### 🤔 Question 1. 
How is the document frequency df of a term used to scale its weight?   
Denoting as usual the total number of documents in a collection by N, we define "INVERSE DOCUMENT" the inverse document frequency (idf) of a term t as follows:  
 - (6.7) $\text{idf}_{t} = \log( \frac{N}{\text{df}_{t}} )$    
Thus the idf of a rare term is high, whereas the idf of a frequent term is likely to be low.  

#### Tf-idf weighting
We now combine the definitions of term frequency and inverse document frequency, to produce a composite weight for each term in each document.
The tf-idf weighting scheme assigns to term t a weight in document d given by  
 - (6.8) $\text{tf-idf}_{t,d} = \text{tf}_{t,d} \times \text{idf}_{t} $   

In other words, tf-idft,d assigns to term t a weight in document d that is
1. highest when t occurs many times within a small number of documents (thus lending high discriminating power to those documents);
2. lower when the term occurs fewer times in a document, or occurs in many documents (thus offering a less pronounced relevance signal);
3. lowest when the term occurs in virtually all documents.  

DOCUMENT VECTOR At this point, we may view each document as a vector with one component corresponding to each term in the dictionary, together with a weight for each component that is given by (6.8).   
For dictionary terms that do not occur in a document, this weight is zero.   
This vector form will prove to be crucial to scoring and ranking.  
As a first step, we introduce the overlap score measure: the score of a document d is the sum, over all query terms, of the number of times each of the query terms occurs in d.   
We can refine this idea so that we add up not the number of  occurrences of each query term t in d, but instead the tf-idf weight of each term in d.
 - (6.9) $Score(q,d) = \sum_{ t \in q} { \text{tf-idf}_{t,d}}$  

#### ✅ 1-1. How does the base of the logarithm in (6.7) affect the score calculation in (6.9)?  

### Effect of the Logarithm Base on Score Calculation  

The base of the logarithm in the inverse document frequency (idf) formula:  

$
\text{idf}_t = \log_b \left( \frac{N}{\text{df}_t} \right)
$

affects the weight assigned to terms and, consequently, the **score** of a document given by:

$
Score(q,d) = \sum_{t \in q} \text{tf-idf}_{t,d}
$

where $\text{tf-idf}_{t,d} = \text{tf}_{t,d} \times \text{idf}_t$.

#### **1. Effect on Score Magnitude**  
Changing the base of the logarithm affects the absolute values of idf and thus the score $ Score(q,d) $.   
Specifically, if we switch from base $ b_1 $ to base $ b_2 $, we use the logarithmic identity:

$
\log_{b_2}(x) = \frac{\log_{b_1}(x)}{\log_{b_1}(b_2)}
$


This means that changing the logarithm base effectively **scales all idf values** by a constant factor. Therefore, the final document scores will also be **scaled by the same constant factor**.


For example, if we switch from natural logarithm ($\ln$, base $ e $) to base 10:

$
\log_{10} x = \frac{\ln x}{\ln 10} \approx 0.4343 \ln x
$

So, all scores computed using base 10 would be about **43% of the scores computed using natural logarithm**, but the ranking order remains the same.

在計算文檔權重時，對數基數的選擇（公式6.7）會影響最終tf-idf分數的計算（公式6.9），但不會改變文檔的相對排名。

在逆向文檔頻率(idf)公式中：
$\text{idf}_{t} = \log(\frac{N}{\text{df}_{t}})$

改變對數的基數會對idf值產生標量效應：
- 如果使用自然對數（基數e），idf值會相對較小
- 如果使用基數10的對數，idf值會相對較大
- 如果使用基數2的對數，則對應於「每次文檔頻率減半，增加一個固定權重」的語義

這個標量效應會直接影響公式6.9的分數計算：
$Score(q,d) = \sum_{t \in q}{\text{tf-idf}_{t,d}}$

因為tf-idf是tf與idf的乘積，所以對數基數的選擇會影響每個查詢詞對最終分數的貢獻程度。然而，由於這種影響是整個系統中一致的標量變化，它不會改變文檔之間的相對排名。換句話說，無論選擇哪種對數基數，如果文檔A的分數高於文檔B，在更改對數基數後，A仍然會排在B之前。

在實踐中，對數基數的選擇通常取決於系統設計和實際應用需求，沒有絕對的「最佳」基數。

$
\text{idf}_t = \log_b \left( \frac{N}{\text{df}_t} \right) = \frac{\log_{10} \left( \frac{N}{\text{df}_t} \right)}{\log_{10} \left( b \right)}
$  
🗣️結論：更改底数從10為b, 会使分数乘以一个因子 $\frac{1}{\log_{10}(b)}$，但给定查询下两个文档的相对分数不受影响。

#### ✅ 1-2. How does the base of the logarithm affect the relative scores of two documents on a given query?  

#### **2. Effect on Relative Scores of Two Documents**  
Since all idf values (and hence tf-idf weights) are scaled **proportionally**, the relative ranking of two documents remains unchanged. That is:

- If $ Score(q, d_1) > Score(q, d_2) $ using one base, the same inequality holds for any other logarithm base.
- The **relative differences** between document scores remain consistent across different logarithm bases.

#### **Conclusion**  
- The choice of the logarithm base affects the absolute magnitude of document scores but **not the ranking** of documents.
- In practice, bases such as 2, 10, or $ e $ are commonly used, but they do not impact the effectiveness of ranking, only the scale of scores.

對數基數在idf計算中（公式6.7）不會影響兩個文檔對於給定查詢的相對分數。

當比較兩個文檔d₁和d₂對於同一查詢q的分數時：

對於文檔d₁：
$Score(q,d₁) = \sum_{t \in q}{\text{tf}_{t,d₁} \times \log_b(\frac{N}{\text{df}_{t}})}$

對於文檔d₂：
$Score(q,d₂) = \sum_{t \in q}{\text{tf}_{t,d₂} \times \log_b(\frac{N}{\text{df}_{t}})}$

如果我們將對數基數從b更改為另一個基數b'，這相當於將所有idf值乘以一個常數因子（log_b'(x) = log_b(x) × log_b'(b)）：

$\log_{b'}(\frac{N}{\text{df}_{t}}) = \log_b(\frac{N}{\text{df}_{t}}) \times \log_{b'}(b)$

由於這個常數因子log_{b'}(b)平等地應用於所有文檔中的所有詞項，它實際上是將所有文檔分數按相同的常數進行縮放：

$Score'(q,d) = Score(q,d) \times \log_{b'}(b)$

因此，如果使用基數b時Score(q,d₁) > Score(q,d₂)，那麼：
$Score'(q,d₁) = Score(q,d₁) \times \log_{b'}(b) > Score(q,d₂) \times \log_{b'}(b) = Score'(q,d₂)$

無論選擇何種對數基數，文檔的相對排名都保持不變。這就是為什麼在實踐中，不同的信息檢索系統可能使用不同的對數基數（例如，自然對數、基數2的對數或基數10的對數）而不影響文檔排名結果。

### 🤔 Question 2. SVD(singular value decomposition)
Let C be the term-document incidence matrix for a collection.  
The number of linearly independent rows in C is 2 ==> rank(C) = 2.

In [17]:
import numpy as np

# Given matrix C
C = np.array([[1, 1],
              [0, 1],
              [1, 0]])

Given that the SVD of the matrix C is U * Σ * V^T.

In [18]:
# Given SVD components
U = np.array([[-0.816, 0.000],
              [-0.408, -0.707],
              [-0.408, 0.707]])

Sigma = np.array([[1.732, 0.000],
                  [0.000, 1.000]])

VT = np.array([[-0.707, -0.707],
               [ 0.707, -0.707]])

In [19]:
# Compute C * C^T and C^T * C
CCT = C @ C.T
CTC = C.T @ C

In [20]:
# Extract the first column
u_col_1 = U[:, 0]
print("First column:\n", u_col_1)

# Extract the second column
u_col_2 = U[:, 1]
print("\nSecond column:\n", u_col_2)

result_1 = np.dot(CCT, u_col_1)
print("CCT * u_col_1:\n", result_1)

#calculating the eigenvalue.
eigenvalue_1 = result_1[0] / u_col_1[0]
print("Eigenvalue 1:", eigenvalue_1)

print()
result_2 = np.dot(CCT, u_col_2)
print("CCT * u_col_2:\n", result_2)

eigenvalue_2 = result_2[1] / u_col_2[1]
print("Eigenvalue 2:", eigenvalue_2)

First column:
 [-0.816 -0.408 -0.408]

Second column:
 [ 0.    -0.707  0.707]
CCT * u_col_1:
 [-2.448 -1.224 -1.224]
Eigenvalue 1: 3.0

CCT * u_col_2:
 [ 0.    -0.707  0.707]
Eigenvalue 2: 1.0


| Column Name | Values                 | CCT * Column | Eigenvalue |
|-------------|------------------------|--------------|------------|
| First Column | [-0.816 -0.408 -0.408] | [-2.448 -1.224 -1.224] | 3.0        |
| Second Column| [ 0.    -0.707  0.707] | [ 0.    -0.707  0.707] | 1.0        |

#### 驗證投影片 vsmodel_2025 第 60 頁的結果
From the python code we can see that 
 - the 1st column of U are orthogonal eigenvector corresponding to the eigenvalue 3.0 of C * C^T 
 - the 2nd column of U are orthogonal eigenvector corresponding to the eigenvalue 1.0 of C * C^T 
```python
CCT  = array([  [2, 1, 1],
                [1, 1, 0],
                [1, 0, 1]])
```
```python
u_col_1 = np.array([[-0.816],
                    [-0.408],
                    [-0.408]]) 
u_col_2 = np.array([ [0.000],
                    [-0.707],
                     [0.707]]) 
```

In [25]:
# V is the matrix of eigenvectors (columns are eigenvectors)
V = VT.T  # Transpose VT to get V

# Extract columns from V
v_col_1 = V[:, 0]
v_col_2 = V[:, 1]

print("First column of V:\n", v_col_1)
print("\nSecond column of V:\n", v_col_2)

# Calculate CTC * v_col_1 and CTC * v_col_2
result_1 = np.dot(CTC, v_col_1)
result_2 = np.dot(CTC, v_col_2)

print("\nCTC * v_col_1:\n", result_1)
print("\nCTC * v_col_2:\n", result_2)

# Calculate eigenvalues (should match np.linalg.eig)
eigenvalue_1 = result_1[0] / v_col_1[0]
eigenvalue_2 = result_2[0] / v_col_2[0]

print("\nEigenvalue 1:", eigenvalue_1)
print("Eigenvalue 2:", eigenvalue_2)

# Create markdown table
markdown_table = """
| Column Name | Values                 | CTC * Column | Eigenvalue |
|-------------|------------------------|--------------|------------|
| First Column | {} | {} | {} |
| Second Column| {} | {} | {} |
""".format(v_col_1, result_1, eigenvalue_1, v_col_2, result_2, eigenvalue_2)

print(markdown_table)

First column of V:
 [-0.707 -0.707]

Second column of V:
 [ 0.707 -0.707]

CTC * v_col_1:
 [-2.121 -2.121]

CTC * v_col_2:
 [ 0.707 -0.707]

Eigenvalue 1: 3.0
Eigenvalue 2: 1.0

| Column Name | Values                 | CTC * Column | Eigenvalue |
|-------------|------------------------|--------------|------------|
| First Column | [-0.707 -0.707] | [-2.121 -2.121] | 3.0 |
| Second Column| [ 0.707 -0.707] | [ 0.707 -0.707] | 1.0 |



| Column Name | Values                 | CTC * Column | Eigenvalue |
|-------------|------------------------|--------------|------------|
| First Column | [-0.707 -0.707] | [-2.121 -2.121] | 3.0 |
| Second Column| [ 0.707 -0.707] | [ 0.707 -0.707] | 1.0 |

#### 驗證投影片 vsmodel_2025 第 60 頁的結果
From the python code we can see that 
 - the 1st column of V are orthogonal eigenvector corresponding to the eigenvalue 3.0 of C^T * C 
 - the 2nd column of V are orthogonal eigenvector corresponding to the eigenvalue 1.0 of C^T * C 

```python
CTC  = array([  [2, 1],
                [1, 2]])
```
```python
v_col_1 = np.array([[-0.707],
                    [-0.707]]) 
v_col_2 = np.array([ [0.707],
                    [-0.707]]) 
```

#### ✅ 2-a. Show the first two largest eigenvalues of ( C * C^T ) are the same as those of ( C^T * C ).

#### ✅ 2-a-1. calculate the eigenvalues and eigenvectors of ( C * C^T )

## Eigenvalues and Eigenvectors of CCT

Given the matrix:

```
CCT = [[2, 1, 1],
       [1, 1, 0],
       [1, 0, 1]]
```

**1. Calculate Eigenvalues**

The characteristic equation is `det(CCT - λI) = 0`, where `λ` are the eigenvalues and `I` is the identity matrix.

$$
CCT - \lambda I = \begin{bmatrix} 2 - \lambda & 1 & 1 \\ 1 & 1 - \lambda & 0 \\ 1 & 0 & 1 - \lambda \end{bmatrix}
$$

The determinant is:

$$
\det(CCT - \lambda I) = (2 - \lambda)[(1 - \lambda)^2 - 0] - 1[(1 - \lambda) - 0] + 1[0 - (1 - \lambda)] \\
= (2 - \lambda)(1 - 2\lambda + \lambda^2) - (1 - \lambda) - (1 - \lambda) \\
= 2 - 4\lambda + 2\lambda^2 - \lambda + 2\lambda^2 - \lambda^3 - 2 + 2\lambda \\
= -\lambda^3 + 4\lambda^2 - 3\lambda \\
= -\lambda(\lambda^2 - 4\lambda + 3) \\
= -\lambda(\lambda - 3)(\lambda - 1)
$$

Setting the determinant to zero:

$$
-\lambda(\lambda - 3)(\lambda - 1) = 0
$$

Thus, the eigenvalues are:

$$
\lambda_1 = 3 \\
\lambda_2 = 1 \\
\lambda_3 = 0
$$

**2. Calculate Eigenvectors**

* **For λ₁ = 3:**

    Substitute λ₁ into `(CCT - λI)v = 0`:

    $$
    \begin{bmatrix} -1 & 1 & 1 \\ 1 & -2 & 0 \\ 1 & 0 & -2 \end{bmatrix} \begin{bmatrix} x \\ y \\ z \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \\ 0 \end{bmatrix}
    $$

    Solving this system of equations, we get:

    $$
    v_1 = \begin{bmatrix} -0.816 \\ -0.408 \\ -0.408 \end{bmatrix}
    $$

* **For λ₂ = 1:**

    Substitute λ₂ into `(CCT - λI)v = 0`:

    $$
    \begin{bmatrix} 1 & 1 & 1 \\ 1 & 0 & 0 \\ 1 & 0 & 0 \end{bmatrix} \begin{bmatrix} x \\ y \\ z \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \\ 0 \end{bmatrix}
    $$

    Solving this system of equations, we get:

    $$
    v_2 = \begin{bmatrix} 0 \\ -0.707 \\ 0.707 \end{bmatrix}
    $$

* **For λ₃ = 0:**

    Substitute λ₃ into `(CCT - λI)v = 0`:

    $$
    \begin{bmatrix} 2 & 1 & 1 \\ 1 & 1 & 0 \\ 1 & 0 & 1 \end{bmatrix} \begin{bmatrix} x \\ y \\ z \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \\ 0 \end{bmatrix}
    $$

    Solving this system of equations, we get:

    $$
    v_3 = \begin{bmatrix} 0.577 \\ -0.577 \\ -0.577 \end{bmatrix}
    $$

**Summary**

* Eigenvalues:
    $$
    \lambda_1 = 3 \\
    \lambda_2 = 1 \\
    \lambda_3 = 0
    $$
* Eigenvectors:
    $$
    v_1 \approx \begin{bmatrix} -0.816 \\ -0.408 \\ -0.408 \end{bmatrix} \\
    v_2 \approx \begin{bmatrix} 0 \\ -0.707 \\ 0.707 \end{bmatrix} \\
    v_3 \approx \begin{bmatrix} 0.577 \\ -0.577 \\ -0.577 \end{bmatrix}
    $$

#### ✅ 2-a-2. calculate the eigenvalues and eigenvectors of ( C^T * C )

## Eigenvalues and Eigenvectors of CTC

Given the matrix:

```
CTC = [[2, 1],
       [1, 2]]
```

**1. Calculate Eigenvalues**

The characteristic equation is `det(CTC - λI) = 0`, where `λ` are the eigenvalues and `I` is the identity matrix.

$$
CTC - \lambda I = \begin{bmatrix} 2 - \lambda & 1 \\ 1 & 2 - \lambda \end{bmatrix}
$$

The determinant is:

$$
\det(CTC - \lambda I) = (2 - \lambda)(2 - \lambda) - (1)(1) = \lambda^2 - 4\lambda + 4 - 1 = \lambda^2 - 4\lambda + 3
$$

Setting the determinant to zero:

$$
\lambda^2 - 4\lambda + 3 = 0 \\
(\lambda - 3)(\lambda - 1) = 0
$$

Thus, the eigenvalues are:

$$
\lambda_1 = 3 \\
\lambda_2 = 1
$$

**2. Calculate Eigenvectors**

* **For λ₁ = 3:**

    Substitute λ₁ into `(CTC - λI)v = 0`:

    $$
    \begin{bmatrix} 2 - 3 & 1 \\ 1 & 2 - 3 \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \end{bmatrix} \\
    \begin{bmatrix} -1 & 1 \\ 1 & -1 \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \end{bmatrix}
    $$

    This gives us the equations:

    $$
    -x + y = 0 \\
    x - y = 0
    $$

    Both equations simplify to `x = y`. Let `x = 1`, then `y = 1`. So, the eigenvector is `v₁ = [1, 1]`.

    Normalize the eigenvector:

    $$
    ||v_1|| = \sqrt{1^2 + 1^2} = \sqrt{2} \\
    v_{1,\text{normalized}} = \begin{bmatrix} 1/\sqrt{2} \\ 1/\sqrt{2} \end{bmatrix} \approx \begin{bmatrix} 0.707 \\ 0.707 \end{bmatrix}
    $$

* **For λ₂ = 1:**

    Substitute λ₂ into `(CTC - λI)v = 0`:

    $$
    \begin{bmatrix} 2 - 1 & 1 \\ 1 & 2 - 1 \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \end{bmatrix} \\
    \begin{bmatrix} 1 & 1 \\ 1 & 1 \end{bmatrix} \begin{bmatrix} x \\ y \end{bmatrix} = \begin{bmatrix} 0 \\ 0 \end{bmatrix}
    $$

    This gives us the equation:

    $$
    x + y = 0
    $$

    Let `x = 1`, then `y = -1`. So, the eigenvector is `v₂ = [1, -1]`.

    Normalize the eigenvector:

    $$
    ||v_2|| = \sqrt{1^2 + (-1)^2} = \sqrt{2} \\
    v_{2,\text{normalized}} = \begin{bmatrix} 1/\sqrt{2} \\ -1/\sqrt{2} \end{bmatrix} \approx \begin{bmatrix} 0.707 \\ -0.707 \end{bmatrix}
    $$

**Summary**

* Eigenvalues:
    $$
    \lambda_1 = 3 \\
    \lambda_2 = 1
    $$
* Eigenvectors:
    $$
    v_1 \approx \begin{bmatrix} 0.707 \\ 0.707 \end{bmatrix} \\
    v_2 \approx \begin{bmatrix} 0.707 \\ -0.707 \end{bmatrix}
    $$

#### ✅ 2-a-3。顯示（ C * C^T ）的前兩個最大特徵值與（ C^​​T * C ）的前兩個最大特徵值相同。
* Eigenvalues of $C * C^T$:
    $$
    \lambda_1 = 3 \\
    \lambda_2 = 1 \\
    \lambda_3 = 0
    $$

* Eigenvalues of $C^T * C$:
    $$
    \lambda_1 = 3 \\
    \lambda_2 = 1
    $$

1️⃣透過數值推導，我們已經證明了（C * C^T）的前兩個最大特徵值與（C^T * C）的前兩個最大特徵值相同。    
對於我們的例子，從計算結果可以看出：
- C * C^T 的特徵值（按降序排列）是 3, 1, 0
- C^T * C 的特徵值是 3, 1

C * C^T 是 3×3 矩陣，有 3 個特徵值，而 C^T * C 是 2×2 矩陣，只有 2 個特徵值。

由於矩陣 C 的秩最多為 2（因為它是 3×2 矩陣），所以 C * C^T 的一個特徵值必然為 0。除此之外，兩個矩陣的非零特徵值是完全相同的。

因此，證明了 C * C^T 和 C^T * C 的最大兩個特徵值確實相同，分別是 3 和 1。這些值與我們從奇異值計算出的值（σ₁² = 1.732² ≈ 3，σ₂² = 1）一致。

2️⃣我們用Python再來證明一下。

In [22]:
# Compute eigenvalues
eigvals_CCT = np.linalg.eigvals(CCT)
eigvals_CTC = np.linalg.eigvals(CTC)

eigvals_CCT_sorted = np.sort(eigvals_CCT)[::-1]
eigvals_CTC_sorted = np.sort(eigvals_CTC)[::-1]

# Print results
print("Largest two eigenvalues of C * C^T:", eigvals_CCT_sorted[:2])
print("Largest two eigenvalues of C^T * C:", eigvals_CTC_sorted[:2])


Largest two eigenvalues of C * C^T: [3. 1.]
Largest two eigenvalues of C^T * C: [3. 1.]


3️⃣第三種方法，我們用理論推導來證明這一點。

# SVD: Singular Value Decomposition

A matrix $C$ can be decomposed as:

$$
C = U \Sigma V^T
$$

where:

- $U$ is an orthogonal matrix  
- $\Sigma$ is a diagonal matrix, with singular values of $C$ on the diagonal  
- $V^T$ is the transpose of $V$, and $V$ is also an orthogonal matrix  

## Computing $C C^T$

$$
C C^T = (U \Sigma V^T) (V \Sigma^T U^T)
$$

$$
= U \Sigma (V^T V) \Sigma^T U^T
$$

Since $V$ is an orthogonal matrix, we have $V^T V = I$ (the identity matrix),

$$
C C^T = U \Sigma I \Sigma^T U^T
$$

$$
= U \Sigma \Sigma^T U^T
$$

Since $\Sigma$ is a diagonal matrix, $\Sigma \Sigma^T$ results in a diagonal matrix with squared singular values on the diagonal.

## Computing $C^T C$

$$
C^T C = (V \Sigma^T U^T) (U \Sigma V^T)
$$

$$
= V \Sigma^T (U^T U) \Sigma V^T
$$

Since $U$ is an orthogonal matrix, we have $U^T U = I$ (the identity matrix),

$$
C^T C = V \Sigma^T I \Sigma V^T
$$

$$
= V \Sigma^T \Sigma V^T
$$

$$
= V (\Sigma^T \Sigma) V^T
$$

## Relationship of Eigenvalues

From our calculations, we can see that:

- $$ C C^T = U (\Sigma \Sigma^T) U^T $$
- $$ C^T C = V (\Sigma^T \Sigma) V^T $$

Since $\Sigma$ is a diagonal matrix, the nonzero diagonal elements of $\Sigma \Sigma^T$ and $\Sigma^T \Sigma$ are the same, which are the squares of the singular values.


#### ✅ 2-b-1. Compute a rank 1 approximation C_1 to the matrix C.

In [23]:
# Compute rank-1 approximation C_1
rank = 1
U1 = U[:, :rank]
S1 = Sigma[:rank, :rank]
VT1 = VT[:rank, :]
C1 = U1 @ S1 @ VT1

print("Rank-1 Approximation C_1:\n", C1)

Rank-1 Approximation C_1:
 [[0.99921158 0.99921158]
 [0.49960579 0.49960579]
 [0.49960579 0.49960579]]


## 計算矩陣 C 的秩一近似 C₁

根據奇異值分解定理，矩陣 C 的秩一近似 C₁ 可以使用最大奇異值及其對應的奇異向量來計算：

$$
C_1 = \sigma_1 \times u_1 \times v_1^T
$$

其中：

- $\sigma_1$ 是最大奇異值（從給定的 Sigma 矩陣中為 1.732）
- $u_1$ 是 U 的第一列（對應於最大奇異值）
- $v_1^T$ 是 $V^T$ 的第一行

從給定的 SVD 分解組件：

- $u_1 = \begin{bmatrix} -0.816 \\ -0.408 \\ -0.408 \end{bmatrix}$
- $\sigma_1 = 1.732$
- $v_1^T = \begin{bmatrix} -0.707 & -0.707 \end{bmatrix}$

計算秩一近似：

$$
C_1 = 1.732 \times \begin{bmatrix} -0.816 \\ -0.408 \\ -0.408 \end{bmatrix} \times \begin{bmatrix} -0.707 & -0.707 \end{bmatrix}
$$

得到結果：

$$
C_1 = \begin{bmatrix} 0.999 & 0.999 \\ 0.500 & 0.500 \\ 0.500 & 0.500 \end{bmatrix}
$$

這是原始矩陣 C 的秩一近似。此近似捕捉了數據中的主要模式，同時將維度降低到秩一。

近似誤差的 Frobenius 範數（$\|C - C_1\|_F$）約為 1.0，表示在此近似中損失的信息量。

#### ✅ 2-b-2. What is the Frobenius norm of the error of this approximation?

In [24]:
# Compute Frobenius norm of the error
error_norm = np.linalg.norm(C - C1, 'fro')
print("Frobenius norm of the error:", error_norm)

Frobenius norm of the error: 1.000000932399249


## 計算秩一近似 C₁ 的誤差

要計算秩一近似 C₁ 的誤差，我們需要計算原始矩陣 C 與近似矩陣 C₁ 之間差異的 Frobenius 範數。

Frobenius 範數定義為矩陣所有元素平方和的平方根：

$$
\|E\|_F = \sqrt{\sum_{i}\sum_{j}|e_{ij}|^2}
$$

其中 $E = C - C_1$ 是誤差矩陣。

從計算結果來看，秩一近似 C₁ 的誤差 Frobenius 範數為：

$$
\|C - C_1\|_F \approx 1.0000009324
$$

這個結果非常接近 1.0，這與 SVD 理論是一致的。根據 SVD 理論，當我們使用最大的 k 個奇異值構建近似時，近似誤差的 Frobenius 範數等於被省略的奇異值的平方和的平方根。

在這個例子中，我們只使用了最大的奇異值 $\sigma_1 = 1.732$ 來構建秩一近似，省略了第二個奇異值 $\sigma_2 = 1.0$。因此，理論上 ( page 65 of the slide vsmodel_2025.pdf )的誤差範數應該等於：

$$
\|C - C_1\|_F = \sqrt{\sigma_2^2} = \sqrt{1^2} = 1.0
$$

我們的計算結果與理論預測非常接近，確認了我們的秩一近似是正確的，誤差 Frobenius 範數為 1.0。